### PR TITLE
ASoC: Intel: boards: tgl_sdw_rt5682: Add machine driver

### PR DESCRIPTION
**Pull request by naveen-manohar**
TGL machine driver with RT5682 connected SNDW Link0 with HDMI Audio
& DMIC Support

Signed-off-by: Naveen Manohar <naveen.m@intel.com>

### CODE REVIEW

**Comment by naveen-manohar** [comment]
Patch for Debug Purposes Only

**Comment by kv2019i** [comment]
@naveen-manohar wrote:
> Patch for Debug Purposes Only

Ack, I added "Draft-for-testing-not-ready-for-review" to inform people to not review this. Please remove the tag if you want to get reviews.

**Comment by plbossart** [comment]
@naveen-manohar could we look into reusing the same driver as for other SoundWire platforms? 
The SOF driver will provide information telling the sdw_rt711_rt1308_rt715.c driver that there is an rt5682 on link1, so in theory all you need to do is provide a descriptor for that platform, and all the rest comes for free.

@RanderWang @bardliao now that I think of it, could we support the rt700 with the same machine driver, instead of maintaining sdw_rt700.c. That way we have a single machine driver for all SOF/SDW platforms.

**Comment by RanderWang** [comment]
> @naveen-manohar could we look into reusing the same driver as for other SoundWire platforms?
> The SOF driver will provide information telling the sdw_rt711_rt1308_rt715.c driver that there is an rt5682 on link1, so in theory all you need to do is provide a descriptor for that platform, and all the rest comes for free.
> 
> @RanderWang @bardliao now that I think of it, could we support the rt700 with the same machine driver, instead of maintaining sdw_rt700.c. That way we have a single machine driver for all SOF/SDW platforms.

yes,  the sdw_rt711_rt1308_rt715 can support rt700 and 5682 or other sdw codecs with codec info added.

File: sound/soc/intel/boards/Kconfig [hunk 1]
```diff
@@ -580,4 +580,22 @@ config SND_SOC_INTEL_SOUNDWIRE_RT711_RT1308_RT715_MACH
 endif
 
 
+if SND_SOC_SOF_TIGERLAKE && SND_SOC_SOF_INTEL_SOUNDWIRE
+
+config SND_SOC_INTEL_TGL_SOUNDWIRE_RT5682_MACH
+       tristate "TGL with RT5682 in SDW mode"
+	depends on I2C && ACPI
```
Comment (lyakh) [inline sound/soc/intel/boards/Kconfig:587]: alignment

File: sound/soc/intel/boards/tgl_sdw_rt5682.c [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+	jack = &rt5682_headset;
+
+	snd_jack_set_key(jack->jack, SND_JACK_BTN_0, KEY_VOLUMEUP);
+	snd_jack_set_key(jack->jack, SND_JACK_BTN_1, KEY_PLAYPAUSE);
+	snd_jack_set_key(jack->jack, SND_JACK_BTN_2, KEY_VOLUMEDOWN);
+	snd_jack_set_key(jack->jack, SND_JACK_BTN_3, KEY_PLAYPAUSE);
+	ret = snd_soc_component_set_jack(component, jack, NULL);
+
+	if (ret == -EAGAIN) {
+		msleep(500);
```
Comment (lyakh) [inline sound/soc/intel/boards/tgl_sdw_rt5682.c:157]: `usleep_range()`

File: sound/soc/intel/boards/tgl_sdw_rt5682.c [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+		ret = snd_soc_component_set_jack(component, jack, NULL);
+	}
+
+	if (ret) {
+		dev_err(rtd->dev, "Headset Jack call-back failed: %d\n", ret);
+		return ret;
```
Comment (lyakh) [inline sound/soc/intel/boards/tgl_sdw_rt5682.c:163]: just do
```
if (ret)
	dev_err(...);

return ret;
```

**Comment by xiulipan** [comment]
@naveen-manohar May I ask what would be the soc-acpi-intel-tgl-match.c change? Can you give an example

**Comment by naveen-manohar** [comment]
> @naveen-manohar could we look into reusing the same driver as for other SoundWire platforms?
> The SOF driver will provide information telling the sdw_rt711_rt1308_rt715.c driver that there is an rt5682 on link1, so in theory all you need to do is provide a descriptor for that platform, and all the rest comes for free.
> 
> @RanderWang @bardliao now that I think of it, could we support the rt700 with the same machine driver, instead of maintaining sdw_rt700.c. That way we have a single machine driver for all SOF/SDW platforms.

@plbossart : Agreed, Let me try re-using sdw_rt711_rt1308_rt715.c for rt5682 at link0.. Its good thought to use single machine driver.

**Comment by naveen-manohar** [comment]
> @naveen-manohar May I ask what would be the soc-acpi-intel-tgl-match.c change? Can you give an example

```
sound/soc/intel/common/soc-acpi-intel-tgl-match.c 
@@ -76,6 +76,12 @@ struct snd_soc_acpi_mach snd_soc_acpi_intel_tgl_sdw_machines[] = {
		.sof_fw_filename = "sof-tgl.ri",
		.sof_tplg_filename = "sof-tgl-rt711-rt1308.tplg",
	},
	{
		.link_mask = 0x1, /* this will only enable rt5682 for now */
		.drv_name = "tgl_sdw_rt5682",
		.sof_fw_filename = "sof-tgl.ri",
		.sof_tplg_filename = "sof-tgl-rt5682.tplg",
	},
```

**Comment by RanderWang** [comment]
> > @naveen-manohar May I ask what would be the soc-acpi-intel-tgl-match.c change? Can you give an example
> 
> ```
> sound/soc/intel/common/soc-acpi-intel-tgl-match.c 
> @@ -76,6 +76,12 @@ struct snd_soc_acpi_mach snd_soc_acpi_intel_tgl_sdw_machines[] = {
> 		.sof_fw_filename = "sof-tgl.ri",
> 		.sof_tplg_filename = "sof-tgl-rt711-rt1308.tplg",
> 	},
> 	{
> 		.link_mask = 0x1, /* this will only enable rt5682 for now */
> 		.drv_name = "tgl_sdw_rt5682",
> 		.sof_fw_filename = "sof-tgl.ri",
> 		.sof_tplg_filename = "sof-tgl-rt5682.tplg",
> 	},
> ```

@naveen-manohar  we add more features in mach table,  can I update your patch and create a PR for you ?

**Comment by naveen-manohar** [comment]
Posted following patches based on PR:1889:
https://github.com/thesofproject/linux/pull/1912/commits/69b6d9b326c38483ac2661f55160e010e634c985
https://github.com/thesofproject/linux/pull/1912/commits/0c92b2853afcb01c60b6b1c18d7df54252d459f3

**Comment by plbossart** [comment]
Replaced by https://github.com/thesofproject/linux/pull/1912

**Pull request closed without merging**